src/main/java/org/folio/dew/repository/AbstractFilesStorage.java

```diff
@@ -0,0 +1,332 @@
+package org.folio.dew.repository;
+
+import org.apache.commons.lang3.StringUtils;
+import org.folio.dew.config.properties.MinioClientProperties;
+import org.folio.s3.client.FolioS3Client;
+import org.folio.s3.client.PutObjectAdditionalOptions;
+import org.folio.s3.client.RemoteStorageWriter;
+import org.folio.s3.client.S3ClientFactory;
+import org.folio.s3.exception.S3ClientException;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.SequenceInputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import io.minio.http.Method;
+import lombok.extern.log4j.Log4j2;
+
+import static org.folio.dew.utils.Constants.PATH_SEPARATOR;
+import static org.folio.dew.utils.Constants.TMP_PATH_PREFIX;
+
+/** Custom extension of FolioS3Client with additional utility methods */
+@Log4j2
+public abstract class AbstractFilesStorage implements FolioS3Client {
+
+  private final FolioS3Client client;
+
+  private final MinioClientProperties properties;
+
+  protected AbstractFilesStorage(MinioClientProperties properties) {
+    this.client = S3ClientFactory.getS3Client(properties.toS3ClientProperties());
+    this.properties = properties;
+
+    this.createBucketIfNotExists();
+  }
+
+  @Override
+  public String upload(String path, String filename) {
+    return client.upload(sanitizePath(path), filename);
+  }
+
+  @Override
+  public String append(String path, InputStream newData) {
+      try {
+          if (!exists(path)) {
+              return write(path, newData);
+          }
+          InputStream existingData = read(path);
+          try (SequenceInputStream combinedStream = new SequenceInputStream(existingData, newData)) {
+              return write(path, combinedStream);
+          }
+      } catch (IOException e) {
+          throw log.throwing(new S3ClientException("Error appending data to file: " + path, e));
+      }
+  }
+
+  public void append(String path, byte[] bytes) {
+    append(path, new ByteArrayInputStream(bytes));
+}
+
+  @Override
+  public String write(String path, InputStream is) {
+    return client.write(sanitizePath(path), is);
+  }
+
+  public String write(String path, byte[] bytes) {
```

Comment:
Is it possible to avoid operations with intermediate byte[] because this potentially leads to memory peaks and OOM as a result?